File: composeApp/src/androidMain/kotlin/com/kissangram/viewmodel/OtherUserProfileViewModel.kt
package com.kissangram.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kissangram.model.Post
import com.kissangram.model.User
import com.kissangram.repository.AndroidAuthRepository
import com.kissangram.repository.AndroidFollowRepository
import com.kissangram.repository.AndroidPreferencesRepository
import com.kissangram.repository.FirestorePostRepository
import com.kissangram.repository.FirestoreUserRepository
import com.kissangram.usecase.FollowUserUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class OtherUserProfileUiState(
    val user: User? = null,
    val posts: List<Post> = emptyList(),
    val isLoading: Boolean = false,
    val isLoadingPosts: Boolean = false,
    val isFollowLoading: Boolean = false,
    val isFollowing: Boolean = false,
    val error: String? = null
)

class OtherUserProfileViewModel(
    application: Application
) : AndroidViewModel(application) {

    private val prefs = AndroidPreferencesRepository(application.applicationContext)
    private val authRepository = AndroidAuthRepository(
        context = application.applicationContext,
        activity = null,
        preferencesRepository = prefs
    )
    private val userRepository = FirestoreUserRepository(authRepository = authRepository)
    private val postRepository = FirestorePostRepository(
        authRepository = authRepository,
        userRepository = userRepository
    )
    private val followRepository = AndroidFollowRepository(
        authRepository = authRepository,
        userRepository = userRepository
    )
    private val followUserUseCase = FollowUserUseCase(followRepository)

    private val _uiState = MutableStateFlow(OtherUserProfileUiState())
    val uiState: StateFlow<OtherUserProfileUiState> = _uiState.asStateFlow()

    /**
     * Load user profile by userId and check follow status
     */
    fun loadUserProfile(userId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            
            try {
                // Load user profile
                val user = userRepository.getUser(userId)
                
                if (user == null) {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = "User not found"
                    )
                    return@launch
                }

                // Check if current user is following this user
                val isFollowing = followRepository.isFollowing(userId)

                _uiState.value = _uiState.value.copy(
                    user = user,
                    isFollowing = isFollowing,
                    isLoading = false
                )

                // Load user's posts
                loadPosts(userId)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load profile"
                )
            }
        }
    }

    /**
     * Toggle follow/unfollow status.
     * Keeps the loader visible (and button disabled) until the cloud function
     * completes, preventing rapid clicks from firing multiple API calls.
     */
    fun toggleFollow() {
        // Guard: ignore clicks while a follow/unfollow is already in flight
        if (_uiState.value.isFollowLoading) return

        val userId = _uiState.value.user?.id ?: return
        val isCurrentlyFollowing = _uiState.value.isFollowing

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isFollowLoading = true, error = null)

            try {
                // Optimistic UI update — keeps isFollowLoading = true so the
                // button stays disabled while the cloud function runs.
                val currentUser = _uiState.value.user
                val newFollowingState = !isCurrentlyFollowing
                val newFollowersCount = if (newFollowingState) {
                    (currentUser?.followersCount ?: 0) + 1
                } else {
                    maxOf(0, (currentUser?.followersCount ?: 0) - 1)
                }

                _uiState.value = _uiState.value.copy(
                    isFollowing = newFollowingState,
                    user = currentUser?.copy(followersCount = newFollowersCount)
                )

                // Perform actual follow/unfollow (cloud function)
                followUserUseCase(userId, isCurrentlyFollowing)

                // Reload user to get the authoritative counts from Firestore
                val updatedUser = userRepository.getUser(userId)
                _uiState.value = _uiState.value.copy(
                    user = updatedUser,
                    isFollowing = followRepository.isFollowing(userId),
                    isFollowLoading = false
                )
            } catch (e: Exception) {
                // Rollback optimistic update on error
                _uiState.value = _uiState.value.copy(
                    isFollowing = isCurrentlyFollowing,
                    isFollowLoading = false,
                    error = e.message ?: "Failed to ${if (isCurrentlyFollowing) "unfollow" else "follow"} user"
                )
            }
        }
    }

    private fun loadPosts(userId: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoadingPosts = true)
            try {
                val posts = postRepository.getPostsByUser(userId, page = 0, pageSize = 30)
                _uiState.value = _uiState.value.copy(posts = posts, isLoadingPosts = false)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoadingPosts = false,
                    posts = emptyList()
                )
            }
        }
    }
}

File: iosApp/iosApp/viewmodel/OtherUserProfileViewModel.swift
import Foundation
import SwiftUI
import Shared

@MainActor
class OtherUserProfileViewModel: ObservableObject {
    @Published var user: User?
    @Published var posts: [Post] = []
    @Published var isLoading = false
    @Published var isLoadingPosts = false
    @Published var isFollowLoading = false
    @Published var isFollowing = false
    @Published var error: String?

    private let prefs = IOSPreferencesRepository()
    private lazy var authRepo = IOSAuthRepository(preferencesRepository: prefs)
    private lazy var userRepo = FirestoreUserRepository(authRepository: authRepo)
    private lazy var postRepo = FirestorePostRepository(authRepository: authRepo, userRepository: userRepo)
    private lazy var followRepo = IOSFollowRepository(authRepository: authRepo, userRepository: userRepo)
    private lazy var followUserUseCase = FollowUserUseCase(followRepository: followRepo)

    func loadUserProfile(userId: String) {
        isLoading = true
        error = nil
        
        Task {
            do {
                // Load user profile
                guard let loadedUser = try await userRepo.getUser(userId: userId) else {
                    await MainActor.run {
                        self.isLoading = false
                        self.error = "User not found"
                    }
                    return
                }

                // Check if current user is following this user
                let followingStatus = try await followRepo.isFollowing(userId: userId)
                
                await MainActor.run {
                    self.user = loadedUser
                    self.isFollowing = followingStatus.boolValue
                    self.isLoading = false
                }
                
                // Load user's posts
                await loadPosts(userId: userId)
            } catch {
                await MainActor.run {
                    self.isLoading = false
                    self.error = error.localizedDescription
                }
            }
        }
    }

    func toggleFollow() {
        guard let userId = user?.id else { return }
        let isCurrentlyFollowing = isFollowing

        isFollowLoading = true
        error = nil

        // Optimistic update
        let currentUser = user
        let newFollowingState = !isCurrentlyFollowing
        let newFollowersCount = if newFollowingState {
            (currentUser?.followersCount ?? 0) + 1
        } else {
            max(0, (currentUser?.followersCount ?? 0) - 1)
        }

        isFollowing = newFollowingState
        if var updatedUser = currentUser {
            updatedUser = User(
                id: updatedUser.id,
                phoneNumber: updatedUser.phoneNumber,
                name: updatedUser.name,
                username: updatedUser.username,
                profileImageUrl: updatedUser.profileImageUrl,
                bio: updatedUser.bio,
                role: updatedUser.role,
                verificationStatus: updatedUser.verificationStatus,
                location: updatedUser.location,
                expertise: updatedUser.expertise,
                followersCount: Int32(newFollowersCount),
                followingCount: updatedUser.followingCount,
                postsCount: updatedUser.postsCount,
                language: updatedUser.language,
                createdAt: updatedUser.createdAt,
                lastActiveAt: updatedUser.lastActiveAt
            )
            user = updatedUser
        }

        Task {
            do {
                // Perform actual follow/unfollow
                try await followUserUseCase.invoke(userId: userId, isCurrentlyFollowing: isCurrentlyFollowing)

                // Reload user to get updated counts
                if let updatedUser = try await userRepo.getUser(userId: userId) {
                    let followingStatus = try await followRepo.isFollowing(userId: userId)
                    
                    await MainActor.run {
                        self.user = updatedUser
                        self.isFollowing = followingStatus.boolValue
                        self.isFollowLoading = false
                    }
                }
            } catch {
                // Rollback optimistic update on error
                await MainActor.run {
                    self.isFollowing = isCurrentlyFollowing
                    self.isFollowLoading = false
                    self.error = "Failed to \(isCurrentlyFollowing ? "unfollow" : "follow") user"
                }
            }
        }
    }
    
    private func loadPosts(userId: String) async {
        await MainActor.run { isLoadingPosts = true }
        do {
            let loadedPosts = try await postRepo.getPostsByUser(userId: userId, page: 0, pageSize: 30)
            await MainActor.run {
                self.posts = loadedPosts
                self.isLoadingPosts = false
            }
        } catch {
            await MainActor.run {
                self.posts = []
                self.isLoadingPosts = false
            }
        }
    }
}
